proxy/src/auth/backend/jwt.rs

```diff
@@ -204,21 +221,32 @@ impl JwkCacheEntryLock {
         ensure!(header.typ == "JWT");
         let kid = header.kid.context("missing key id")?;
 
-        let mut guard = self.get_or_update_jwk_cache(client, fetch).await?;
+        let mut guard = self
+            .get_or_update_jwk_cache(ctx, client, role_name.clone(), fetch)
+            .await?;
 
         // get the key from the JWKs if possible. If not, wait for the keys to update.
-        let jwk = loop {
+        let (jwk, aud) = loop {
             let jwk = guard
                 .key_sets
                 .values()
-                .flat_map(|jwks| &jwks.keys)
-                .find(|jwk| jwk.prm.kid.as_deref() == Some(kid));
+                .flat_map(|key_set| {
```

Comment:
This looks good and I'm sure rustc will turn the nested iterators into something efficient, but I wonder if we can increase readability, by extracting part of this into a `KeySet::find_by_id` method that returns `Option<(jwk,aud)>`.
Or maybe a `SuperKeySet` that provides O(1) lookup by ID across contained sets.
But then, maybe this path is not hot enough to warrant this.

Reply:
That reminds me. I need to filter out the JWKs responses to only have 4 entries at most. With only 4 entries, the iterator search is usually faster than an O(1) hashmap lookup.

Reply:
Moved to some helper methods